Jenkinsfile

```diff
@@ -0,0 +1,599 @@
+  jacoco(
+    execPattern: '**/target/jacoco.exec',
+    classPattern: '**/classes',
+    sourcePattern: '**/src/main/java'
+  )
+}
+
+def notifySlack(status = 'started') {
+  // Notify Slack channel for every build except adhoc executions
+  if (params.ADHOC_BUILD_TYPE != 'BUILD-AND-EXECUTE-TESTS') {
+    // Set the global pipeline scoped environment (this is above each matrix)
+    env.BUILD_STATED_SLACK_NOTIFIED = 'true'
+
+    def buildType = 'Commit'
+    if (params.CI_SCHEDULE != 'DO-NOT-CHANGE-THIS-SELECTION') {
+      buildType = "${params.CI_SCHEDULE.toLowerCase().capitalize()}"
+    }
+
+    def color = 'good' // Green
+    if (status.equalsIgnoreCase('aborted')) {
+      color = '808080' // Grey
+    } else if (status.equalsIgnoreCase('unstable')) {
+      color = 'warning' // Orange
+    } else if (status.equalsIgnoreCase('failed')) {
+      color = 'danger' // Red
+    }
+
+    def message = """Build ${status} for ${env.DRIVER_DISPLAY_NAME} [${buildType}]
+<${env.GITHUB_BRANCH_URL}|${env.BRANCH_NAME}> - <${env.RUN_DISPLAY_URL}|#${env.BUILD_NUMBER}> - <${env.GITHUB_COMMIT_URL}|${env.GIT_SHA}>"""
+    if (!status.equalsIgnoreCase('Started')) {
+      message += """
+${status} after ${currentBuild.durationString - ' and counting'}"""
+    }
+
+    slackSend color: "${color}",
+              channel: "#java-driver-dev-bots",
+              message: "${message}"
+  }
+}
+
+def submitCIMetrics(buildType) {
+  long durationMs = currentBuild.duration
+  long durationSec = durationMs / 1000
+  long nowSec = (currentBuild.startTimeInMillis + durationMs) / 1000
+  def branchNameNoPeriods = env.BRANCH_NAME.replaceAll('\\.', '_')
+  def durationMetric = "okr.ci.java.${env.DRIVER_METRIC_TYPE}.${buildType}.${branchNameNoPeriods} ${durationSec} ${nowSec}"
+
+  timeout(time: 1, unit: 'MINUTES') {
+    withCredentials([string(credentialsId: 'lab-grafana-address', variable: 'LAB_GRAFANA_ADDRESS'),
+                     string(credentialsId: 'lab-grafana-port', variable: 'LAB_GRAFANA_PORT')]) {
+      withEnv(["DURATION_METRIC=${durationMetric}"]) {
+        sh label: 'Send runtime metrics to labgrafana', script: '''#!/bin/bash -lex
+          echo "${DURATION_METRIC}" | nc -q 5 ${LAB_GRAFANA_ADDRESS} ${LAB_GRAFANA_PORT}
+        '''
+      }
+    }
+  }
+}
+
+def describePerCommitStage() {
+  script {
+    currentBuild.displayName = "Per-Commit build"
+    currentBuild.description = 'Per-Commit build and testing of development Apache CassandraⓇ and current DataStax Enterprise against Oracle JDK 8'
+  }
+}
+
+def describeAdhocAndScheduledTestingStage() {
+  script {
+    currentBuild.displayName = "Adhoc testing"
+    if (params.CI_SCHEDULE == 'WEEKNIGHTS') {
+      currentBuild.displayName = "${params.CI_SCHEDULE.toLowerCase().replaceAll('_', ' ').capitalize()} schedule"
+    }
+    currentBuild.description = "Testing ${params.ADHOC_BUILD_AND_EXECUTE_TESTS_SERVER_VERSION} against JDK version ${params.ADHOC_BUILD_AND_EXECUTE_TESTS_JABBA_VERSION}"
```

Comment:
This will actually have the wrong information in the job description for scheduled builds. See [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/30/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/30/) as an example. That job description indicates it was run against Cassandra 2.1 with JDK8, when the parameterized build info below shows it was actually run against 3.11, 4.0, 6.7 and 6.8 using JDK 14. I'll fix this in my next commit.